test/flag_anomaly.py

```diff
@@ -0,0 +1,126 @@
+from ase.io import read, write
+from ase.io.trajectory import Trajectory
+from pymatgen.io.ase import AseAtomsAdaptor
+from pymatgen.core.structure import SiteCollection
+from pymatgen.core.surface import SlabGenerator
+from pymatgen.analysis.local_env import VoronoiNN
+
+def get_connectivity(atoms, multiple=1):
+    """
+    Generate the connectivity of an atoms obj.
+    Args:
+                atoms      An `ase.Atoms` object
+    Returns:
+                matrix     The connectivity matrix of the atoms object.
+    """
+    cutoff = natural_cutoffs(atoms, mult=multiple)
+    neighborList = neighborlist.NeighborList(cutoff, self_interaction=False, bothways=True)
+    neighborList.update(atoms)
+    matrix = neighborlist.get_connectivity_matrix(neighborList.nl).toarray()
+    return matrix
+
+def find_max_movement(init_atoms, final_atoms):
+    '''
+    Given ase.Atoms objects, find the furthest distance that any single atom in
+    a set of atoms traveled (in Angstroms)
+    Args:
+            init_atoms      `ase.Atoms` of the structure in its initial state
+            final_atoms     `ase.Atoms` of the structure in its final state
+    Returns:
+            max_movement    A float indicating the further movement of any single atom
+                            before and after relaxation (in Angstroms)
+    '''
+    # Calculate the distances for each atom
+    distances = final_atoms.positions - init_atoms.positions
+
+    # Reduce the distances in case atoms wrapped around (the minimum image convention)
+    _, movements = find_mic(distances, final_atoms.cell, final_atoms.pbc)
+    max_movement = max(movements)
+    return max_movement
+
+def is_adsorbate_placed_correct(adslab, atoms_tag):
+    """
+    Make sure all adsorbate atoms are connected after placement.
+    False means there is at least one isolated adsorbate atom.
+    Args:
+            adslab             `ase.Atoms` of the structure in its initial state
+            atoms_tag (list)    0=bulk, 1=surface, 2=adsorbate
+    Returns:
+            boolean    If there is any stand alone adsorbate atoms after placement,
+                       return False.
+
+    """
+    adsorbate_idx = [idx for idx, tag in enumerate(self.atoms_tag) if tag==2]
+    connectivity = get_connectivity(self.init_atoms[adsorbate_idx])
+    if np.any(np.sum(connectivity, axis=0)==0):
+        return False
+    return True
+
+class DetectAnomaly:
+    def __init__(self, init_atoms, final_atoms, atoms_tag):
+        """
+        flag anomalies based on initial and final stucture of a relaxation.
+        Args:
+               init_atoms         `ase.Atoms` of the adslab in its initial state
+               final_atoms        `ase.Atoms` of the adslab in its final state
+               atoms_tag (list)    0=bulk, 1=surface, 2=adsorbate
+        """
+        self.init_atoms = init_atoms
+        self.final_atoms = final_atoms
+        self.atoms_tag = atoms_tag
+
+    def _is_adsorbate_dissociated(self):
```

Comment:
Style nit: methods starting with _ imply that they're private, but these methods are meant to be called by the user.

Reply:
updated